### PR TITLE
Add Lwt_retry library

### PR DESCRIPTION
**Pull request by shonfeder**
Adds the `Lwt_retry` library discussed in #1028 as a separate package. I believe all guidance suggested there by @raphael-proust has been incorporated. Thanks for that!

I did not add an `iter` function, since I think this doesn't add any value over using `Lwt_stream.iter` directly.

I've tried to make the code, documentation, and tests consistent with the style I found in other locations in the code base.

### CODE REVIEW

File: lwt_retry.opam
Comment (raphael-proust) [inline lwt_retry.opam:7]: I think you can give credit where credit is due (yourself? some colleagues of yours?)

File: lwt_retry.opam
Comment (raphael-proust) [inline lwt_retry.opam:5]:
```suggestion
  "Raphaël Proust <code@bnwr.net>" "Shon Feder <shon.feder@gmail.com>"
```

File: src/retry/lwt_retry.ml
Comment (raphael-proust) [inline src/retry/lwt_retry.ml:42]: Not sure if I've missed something (shouldn't review in the evening probably) but it seems that you call `f` "one too many time". Like, if the `stop` ref is `true`, you still call `f`, you just don't use the result.

Is this intended?

File: src/retry/lwt_retry.ml
Comment (raphael-proust) [inline src/retry/lwt_retry.ml:52]: nano-optimisation

```suggestion
    | Ok _ as ok -> Lwt.return ok
```

Should perform one fewer allocation.

File: src/retry/lwt_retry.ml
Comment (raphael-proust) [inline src/retry/lwt_retry.ml:55]: Same here

```suggestion
    | Error (_, n) as err ->
      let* () = Lwt_unix.sleep @@ duration n in
      Lwt.return err
```

File: src/retry/lwt_retry.ml
Comment (raphael-proust) [inline src/retry/lwt_retry.ml:59]:
```suggestion
  (* The first attempt is a try, and Retries start counting from n + 1 *)
```

File: src/retry/lwt_retry.ml
Comment (raphael-proust) [inline src/retry/lwt_retry.ml:64]: Have you tried feeding negative numbers of `n`?

I'd recommend adding `if n < 0 then raise (Invalid_argument "Lwt_retry.n_times: n must be positive or null");` at the beginning of the function.

File: src/retry/lwt_retry.mli
Comment (raphael-proust) [inline src/retry/lwt_retry.mli:143]: It's sort of implied but I think it would be good to say it explicitly somewhere: `n_times` forces the (n first elements of the) on-demand stream of attempts.

Streams are one of the most confusing part of Lwt's core library so better be explicit about everything.

**Review by raphael-proust — COMMENTED** [review]
Looks good. Some small nitpick-ish comments and a couple of actual suggestions. Feel free to push back though, I could very well have gotten something wrong.

File: lwt_retry.opam
Comment (shonfeder) [reply lwt_retry.opam:7]: Whoops! Oversight here. Sorry! Thanks for catching this :)

File: src/retry/lwt_retry.ml
Comment (shonfeder) [reply src/retry/lwt_retry.ml]: You are right! Thanks for catching this (embarrassing) mistake. Fixed in fc49af070adc60199866af294607da6342b15737 .

File: src/retry/lwt_retry.ml
Comment (shonfeder) [reply src/retry/lwt_retry.ml]: Fixed in c8b1f3977b306a08ae1c04f5730364c89420fcc5 .

File: src/retry/lwt_retry.ml
Comment (shonfeder) [reply src/retry/lwt_retry.ml]: Fixed in 0cc9b60db7165839c5ed645c8c5463dcbd6b7e39

File: src/retry/lwt_retry.ml
Comment (shonfeder) [reply src/retry/lwt_retry.ml]: Good catch! Fixed in ade04564e371f8b0d4c9fa769ba4a6077b0f2418

File: src/retry/lwt_retry.mli
Comment (shonfeder) [reply src/retry/lwt_retry.mli:143]: LMK if you think 34aefc14bde3d66f1640b3c51684defdec6456a7 addresses this!

**Review by shonfeder — COMMENTED** [review]
Thank you for the careful review! I think I've addressed all the suggestions and corrections.

File: lwt_retry.opam
Comment (shonfeder) [reply lwt_retry.opam:7]: Ah, this is just generated based on the dune-project, which is why I missed.

Comment (shonfeder) [reply lwt_retry.opam:7]: Hmm, this is actually trickier than I'd expect, as dune doesn't support `maintainers` or `authors` fields on `package` stanzas until dune 2.0. But there are several other aspects of the lwt dune config that are incompatible with dune 2.0. Most are trivial so far, but some have to do with the C FFI:

```
File "test/retry/dune", lines 8-9, characters 1-33:
8 |  (action
9 |   (run %{exe:main.exe})))
File "_build/.dune/default/src/unix/dune", line 178, characters 1-43:
178 |  (c_flags -I. (:include unix_c_flags.sexp))
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: 'c_flags' was deleted in version 2.0 of the dune language. Use the
(foreign_stubs ...) field instead.
Had 1 error, waiting for filesystem changes...
```

So I think our options here are:

1. Leave the maintainer/authors package as is, generated from the data in the dune-project file.
2. use an opam.template file for the retry package.
3. Update lwt to use dune 2.0, and put the data in the `package` stanza.

3 seems like the right (tm) way to do this to me. If you agree, I can make that change in a separate PR then rebase this on top. WDYT?

Comment (raphael-proust) [reply lwt_retry.opam:7]: For 3 I need to first check the dependency cone for dune.2+ (minimal ocaml version and other such minimal version constraints). One of the thing is we try to keep maximum compatiblity because we treat Lwt as a kind of foundational library.

Depending on the compatibility, I'd go for either 2 or 3. I can check that some times in the coming week.

File: src/retry/lwt_retry.ml
Comment (raphael-proust) [reply src/retry/lwt_retry.ml]: :+1: nice adding a test too

File: src/retry/lwt_retry.ml
Comment (raphael-proust) [reply src/retry/lwt_retry.ml]: ah I see. Either is fine then.

File: src/retry/lwt_retry.mli
Comment (raphael-proust) [reply src/retry/lwt_retry.mli:143]: LGTM

**Comment by raphael-proust** [comment]
Thanks! I'll have a look at the dune version thing this week

File: lwt_retry.opam
Comment (MisterDA) [reply lwt_retry.opam:7]: Dune doesn't have external dependencies. It requires (at least) ocaml 4.08 and has a fallback through a secondary ocaml compiler for earlier versions. I don't think there's anything to be concerned here with.

Comment (shonfeder) [reply lwt_retry.opam:7]: Thank, @MisterDA . Opened a PR for the update at https://github.com/ocsigen/lwt/pull/1035

**Comment by shonfeder** [comment]
Rebased on #1035 so I could update the package author and maintainer data. I'll need to rebase on master once that is merged in.

The only CI failures remain the FreeBSD and MacOS failures that are occurring in the trunk:

```
Test 'getgrgid and Unix.getgrgid' in suite 'lwt_unix' raised 'Not_found'
Test 'getgrnam and Unix.getgrnam' in suite 'lwt_unix' raised 'Not_found'
```

**Comment by raphael-proust** [comment]
> I'll need to rebase on master once that is merged in

you can go ahead @shonfeder

**Comment by shonfeder** [comment]
Rebased! Should be good for a final review and/or merge! Thanks again for the reviews and guidance along the way, @raphael-proust :)

**Comment by shonfeder** [comment]
Anything else you'd like to see here @raphael-proust ?

File: dune-project
Comment (raphael-proust) [inline dune-project:24]:
```suggestion
  (lwt (>= 5.3.0))))
```

because of `Lwt.Syntax`

File: lwt_retry.opam
Comment (raphael-proust) [inline lwt_retry.opam:15]:
```suggestion
  "lwt" {>= "5.3.0"}
```

matching change

File: test/retry/main.ml
Comment (raphael-proust) [inline test/retry/main.ml:147]:
```suggestion
      let duration _ = 0.01 in
```

This is the bottleneck for the test suite and it doesn't make sense to go for tenth of a seconds order of magniture.

**Review by raphael-proust — CHANGES_REQUESTED** [review]

**Review by raphael-proust — APPROVED** [review]

**Comment by raphael-proust** [comment]
Ready to merge unless the CI shows I messaed up something with my suggested+applied changes

**Comment by shonfeder** [comment]
Thank you for the fixes!

**Comment by raphael-proust** [comment]
Thanks for the contribution!!

**Pull request merged**
